lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/StateMachine/DriverTaskState.cs

```diff
@@ -0,0 +1,145 @@
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskRunning, TaskStateEvent.CompletedTask), TaskState.TaskCompleted },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskRunning, TaskStateEvent.WaitingTaskToClose), TaskState.TaskWaitingForClose },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskRunning, TaskStateEvent.FailedTaskAppError), TaskState.TaskFailedByAppError },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskRunning, TaskStateEvent.FailedTaskSystemError), TaskState.TaskFailedBySystemError },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskRunning, TaskStateEvent.FailedTaskEvaluatorError), TaskState.TaskFailedByEvaluatorFailure },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskRunning, TaskStateEvent.FailedTaskCommunicationError), TaskState.TaskFailedByGroupCommunication },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskWaitingForClose, TaskStateEvent.ClosedTask), TaskState.TaskClosedByDriver },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskWaitingForClose, TaskStateEvent.FailedTaskAppError), TaskState.TaskClosedByDriver },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskWaitingForClose, TaskStateEvent.FailedTaskSystemError), TaskState.TaskClosedByDriver },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskWaitingForClose, TaskStateEvent.FailedTaskEvaluatorError), TaskState.TaskClosedByDriver },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskWaitingForClose, TaskStateEvent.FailedTaskCommunicationError), TaskState.TaskClosedByDriver },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskFailedBySystemError, TaskStateEvent.FailedTaskEvaluatorError), TaskState.TaskFailedByEvaluatorFailure },
+            { new StateTransition<TaskState, TaskStateEvent>(TaskState.TaskFailedByGroupCommunication, TaskStateEvent.FailedTaskEvaluatorError), TaskState.TaskFailedByEvaluatorFailure }
+        });
+
+        /// <summary>
+        /// Final state means that the task is not running and will not attempt to run. 
+        /// When all tasks are in final states, we can proceed to recovery or fail. 
+        /// Transitions between final states are possible if causes of task termination are clarified. For example, we might receive 
+        /// FailedTask event first, then set the task state as TaskFailedBySystemError or TaskFailedByGroupCommunication. And later 
+        /// received FailedEvaluator that indicates the attached task was failed is actually caused by evaluator failure. We can then 
+        /// change the state to TaskFailedByEvaluatorFailure.
+        /// For the task state transition diagram <see href="https://issues.apache.org/jira/browse/REEF-1327"></see>
```

Comment:
Unrelated to the code in this PR, but shouldn't we keep only one copy of design document across all JIRAs, say, in parent fault tolerance JIRA? Otherwise, if we have to update the document (which is likely), we'll have to go through all JIRAs which contain copies of it, which is prone to missing some and leaving them not updated